### ISSUE TITLE
Latest package upgrade behaviour doesn't cope with all debian version formats.

### ISSUE THREAD

**Issue by bogdanap** [chef/chef#6549]
## Description
Our builds recently started failing on apt package upgrades. My best assumption would be that the changes introduced in https://github.com/chef/chef/pull/6428 don't take into account all debian version formats, that could also include characters like `~` and `:` (in addition to `+` and `_`) which make Gem::Version unhappy. 

## Chef Version

13.6.0

## Platform Version
Ubuntu 14.04

## Replication Case

```
apt_repository 'pollinate' do
  uri 'http://ppa.launchpad.net/pollinate/ppa/ubuntu'
  distribution node['lsb']['codename']
  components %w[main]
  key '43732553'
  keyserver 'hkp://ha.pool.sks-keyservers.net'
  retries 2
  retry_delay 30
end

package 'pollinate' do
  action %i[install upgrade]
end
```

This is the complete recipe that currently fails at the upgrade package step:
https://github.com/travis-ci/travis-cookbooks/blob/master/cookbooks/travis_build_environment/recipes/cloud_init.rb#L34

## Client Output
```
    googlecompute:   * apt_package[pollinate] action upgrade
    googlecompute:
    googlecompute:     ================================================================================
    googlecompute:     Error executing action `upgrade` on resource 'apt_package[pollinate]'
    googlecompute:     ================================================================================
    googlecompute:
    googlecompute:     ArgumentError
    googlecompute:     -------------
    googlecompute:     Malformed version number string 4.23-0ubuntu1~14.04
```
Failing build: https://travis-ci.org/travis-infrastructure/packer-build/jobs/295254154

**Comment by onlyhavecans** [comment]
I also have this issue

```
         * apt_package[ruby2.3] action upgrade

           ================================================================================
           Error executing action `upgrade` on resource 'apt_package[ruby2.3]'
           ================================================================================

           ArgumentError
           -------------
           Malformed version number string 2.3.5-1bbox2~trusty1

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/dnsimple-server/recipes/_ruby.rb

            18: package 'ruby2.3' do
            19:   action :upgrade
            20:   notifies :reload, 'ohai[refresh_rubies]', :immediately
            21:   notifies :run, 'execute[reinstall_bundler]', :immediately
            22: end
            23:

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/dnsimple-server/recipes/_ruby.rb:18:in `from_file'

           apt_package("ruby2.3") do
             package_name "ruby2.3"
             action [:upgrade]
             default_guard_interpreter :default
             declared_type :package
             cookbook_name "dnsimple-server"
             recipe_name "_ruby"
           end

           System Info:
           ------------
           chef_version=13.6.0
           platform=ubuntu
           platform_version=14.04
           ruby=ruby 2.4.2p198 (2017-09-14 revision 59899) [x86_64-linux]
           program_name=chef-client worker: ppid=30513;start=23:32:34;
           executable=/opt/chef/bin/chef-client
```

**Comment by moutons** [comment]
Specifically we should support the characters in https://www.debian.org/doc/debian-policy/#s-f-version

**Comment by tas50** [comment]
This is fixed in 13.6.2 by using dpkg to do the version comparison since it knows how to parse the versions the best. We also have a bit of an optimization in that upgrade actions against packages that haven't even been installed will no longer go through the comparison process since it's irrelevant.  I've tested things out within test kitchen using a simplified version of the report @bogdanap, but it would be great if you could pull down 13.6.2 from the current channel and give this a test. Open this back up if there's any issues. We'll see if we can get a build shipped out on Monday.

curl -L https://omnitruck.chef.io/install.sh | sudo bash -s -- -c current

```
-----> Converging <default-ubuntu-1604>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 6.3.1...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Chef Omnibus installation detected (install only if missing)
       Transferring files to <default-ubuntu-1604>
       Starting Chef Client, version 13.6.2
       resolving cookbooks for run list: ["apt_test::default"]
       Synchronizing Cookbooks:
         - apt_test (0.1.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: apt_test::default
         * apt_package[pollinate] action upgrade
           - upgrade package pollinate from 4.15-0ubuntu1 to 4.25-0ubuntu1~16.04.1

       Running handlers:
       Running handlers complete
       Chef Client finished, 1/1 resources updated in 04 seconds
       Finished converging <default-ubuntu-1604> (0m8.93s).
```
